[8.x] nestWheres function for Eloquent Builder

bert-w opened the pull request:
This pull request adds an extra function to the Eloquent query builder to nest all `WHERE` statements of the current query, which solves a shortcoming in relations with `orWhere()` clauses.

### Case description showcasing the current limitation
Say a user can have many accounts through a one-to-many relationship, where an account simply has a `user_id` property on it. However, it appears that there is also functionality for so called "shared" accounts: accounts that have an `is_shared` property set to `1`. This results in the following relation:

```php
// Models/User.php
public function accounts()
{
    return $this->hasMany(Account::class, 'user_id')->orWhere('is_shared', 1);
}
```

Now, this relation works exactly as expected. All accounts that have the `user_id` equal the user's id OR all accounts that have `is_shared=1` will come up in this relation.

However, when chaining additional query elements onto this relation, an issue arises as shown here:
```php
public function activeAccounts()
{
    return $this->accounts()->where('active', 1);
}
```
Which results in the query:
```
select *
from `accounts`
where
    `accounts`.`user_id` = 1 and
    `accounts`.`user_id` is not null or
    `is_shared` = 1 and
    `active` = 1
```
If you look closely, this may return accounts that have `active=0`, namely if the `user_id` matches. This is caused by the missing parentheses of the initial where clause:
```
// Wrong (current)
select * from `accounts` where A AND B OR C AND D
// Correct
select * from `accounts` where ( A AND B OR C ) AND D
```

### Solution
Interestingly enough, there already appears to be a fix for this which is directed solely at Query Scopes, see:
https://github.com/laravel/framework/blob/8.x/src/Illuminate/Database/Eloquent/Builder.php#L1119
(`Database/Eloquent/Builder.php :: callScope()`)

This piece of code appears to nest pre-existing where-conditions in order to apply the new query scope successfully.

However, as far as I can see there is no way to apply this nesting manually while working on a query. This means that in order to fix the described issue above, you MUST make use of a query scope which imho increases complexity. So this pull request introduces a new function that can be chained onto your query to nest all current `where` like so:

```php
// Models/User.php
public function accounts()
{
    return $this->hasMany(Account::class, 'user_id')->orWhere('is_shared', 1)->nestWheres();
}
```
outputs: ```select * from `accounts` where ( A AND B OR C )```

Now, the issue is fixed since the `OR` is bracketed with the main hasMany-where clause.

### Open ideas
- Add a `$whereCount` parameter to `nestWheres()` to only nest the first `n` where clauses (this increases complexity)
- Refactor the Eloquent Builder function `addNewWheresWithinGroup()` since it has some overlap with this problem
- Should this become a part of the main Query Builder class?

bert-w committed:
nestWheres function for Eloquent Builder

bert-w committed:
fix empty

taylorotwell commented:
I think this is definitely something that would belong on the base query builder class - not the Eloquent builder. Also, does it work properly if I do something like this:

```php
->where(function ($query) { $query->where('foo', bar'); })->where(fn...)->nestWheres();
```

bert-w committed:
move functions to query builder, change addNewWheresWithinGroup to nestWheres

bert-w committed:
style fix

bert-w commented:
alright, I moved some stuff around to make it work in the main Query Builder class:

- EloquentBuilder `addNewWheresWithinGroup(QueryBuilder $query, $originalWhereCount)`
  - moved to QueryBuilder as `nestWheres(originalWhereCount = null)`
- EloquentBuilder `groupWhereSliceForScope(QueryBuilder $query, $whereSlice)`
   - moved to QueryBuilder  as `groupWhereSlice($whereSlice)`
- EloquentBuilder `createNestedWhere($whereSlice, $boolean = 'and')`
   - moved to QueryBuilder

I've also added a couple of tests that test a simple and an advanced query with the `nestWheres()` function.

bert-w committed:
stray newline

netpok commented:
These changes make this PR breaking and should target 9.x.

driesvints commented:
Please put the moved methods back at their original position so the diff is more clear at what actually changed.

bert-w commented:
> 
> 
> These changes make this PR breaking and should target 9.x.

im not sure about the exact breaking change policy. FYI: Only `protected` methods have been moved, and 1 `public` method has been added. So things can only break if you make extensions to the Builder in your own code. @driesvints 9.x or 8.x?

> 
> 
> Please put the moved methods back at their original position so the diff is more clear at what actually changed.

The methods you are looking at have been moved from `Eloquent/Builder` to `Query/Builder`, so I'm not sure what you mean by original position. Moving the functions is required to make it work (and it actually cleans up the function signatures since they no longer need `QueryBuilder $query` as a 1st param).

driesvints commented:
@bert-w ah right, sorry.

@netpok can you tell us what changes make this breaking?

netpok commented:
Basically any class which extends the builder breaks:
- if it used the moved methods, since the signature (including name and parameter) list is not the same
- if a method is made public than any child class will break if it overrides the method because public method cannot be overridden with a protected one.

This could also be a problem since the builder class is macroable.

A basically it's not breaking when:
- the method is protected or private on a final class
- the method is private on a non-final class

driesvints commented:
@netpok I don't see any method that got its signature changed and I don't see any protected method that was made public. Can you point out the ones that you mean?

> This could also be a problem since the builder class is macroable.

That's not a problem because we can't control what people do in userland. This is the same argument for us adding new methods on minor/patch releases.

bert-w commented on src/Illuminate/Database/Query/Builder.php:
```diff
@@ -1811,6 +1811,72 @@ protected function addDynamic($segment, $connector, $parameters, $index)
+     * @param  string  $boolean
+     * @return array
+     */
+    protected function createNestedWhere($whereSlice, $boolean = 'and')
```
direct copy from `Eloquent/Builder`

bert-w commented on src/Illuminate/Database/Query/Builder.php:
```diff
@@ -1811,6 +1811,72 @@ protected function addDynamic($segment, $connector, $parameters, $index)
+     * @param  int|null  $originalWhereCount
+     * @return $this
+     */
+    public function nestWheres($originalWhereCount = null)
```
Copy from `Eloquent/Builder` with first parameter removed since we are now in the Query class itself.
+ made public
+ slight logic adjustment for the default case of `$originalWhereCount = null` (nest *all* wheres).

driesvints replied:
This method doesn't exists on `Eloquent/Builder` atm? Can you point out where it is?

bert-w replied:
oops correction: `groupWhereSliceForScope` renamed to `groupWhereSlice` since the original function was aimed at scopes specifically.

bert-w replied:
`addNewWheresWithinGroup(...)` renamed to `nestWheres(...)` to be more explicit

bert-w commented on src/Illuminate/Database/Query/Builder.php:
```diff
@@ -1811,6 +1811,72 @@ protected function addDynamic($segment, $connector, $parameters, $index)
+     * @param  array  $whereSlice
+     * @return void
+     */
+    protected function groupWhereSlice($whereSlice)
```
Copy from `Eloquent/Builder` with first parameter removed since we are now in the Query class itself.

driesvints replied:
This method doesn't exists on `Eloquent/Builder` atm? Can you point out where it is?

bert-w replied:
oops: `groupWhereSliceForScope` renamed to `groupWhereSlice` since it is not about scopes specifically anymore.

driesvints commented:
I see, thanks @bert-w. In that case @netpok is correct and this PR should be sent to 9.x. I didn't realize the removed methods didn't live on the base query builder. Please send this to 9.x instead. Thanks
